### PR TITLE
Cognates

### PR DESCRIPTION
**Pull request by joelgrondman**
Cognate generation and user classification based on similarity of words

### CODE REVIEW

File: wordstats/cognate_info.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            self.candidates.extend(newlist)
+
+    # note: only executable in terminal, reactive to one-key stroke
+    def start_quiz(self):
+        print(self.language_ids)
+        print("y:   whitelist")
+        print("q:   quit")
+        print("other:   blacklist")
+
+        for cognate in self.candidates:
```
Comment (mircealungu) [inline wordstats/cognate_info.py]: we want only those candidates which are not already in a whitelist blacklist, right?

File: wordstats/testquiz.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+distanceMetric = WordDistance.loadConfig("denl", "edit_distance_rules")
+distanceMetric.load_rules("denl", "edit_distance_rules")
+
+cognateFM.apply_distance_metric(german, dutch, distanceMetric.edit_distance_rules)
+cognateFM.save_candidates()
+
+print(cognateFM.candidates)
+cognateFM.start_quiz()
+cognateFM.save_evaluation()
```
Comment (mircealungu) [inline wordstats/testquiz.py]: hmm. 
if for an evaluation there are 3000 candidates, then
the only way the info is saved is if somebody runs the program and provides feedback for all the 3000.
if by mistake the program stops for example, nothing is saved. 
i think the info should be saved as soon as somebody provides feedback on a given cognate. 

this also explains my comment on the start_quiz() method.

File: wordstats/cognate_info.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            table = Table('word_info', Base.metadata, autoload=True, autoload_with=BaseService.engine)
+            words = BaseService.session.query(table).filter(table.c.language_id == self.language_id)
+            words.delete(synchronize_session=False)
+
+        clear_corresponding_entries_in_db(self)
+        for word_info in list(self.word_info_dict.values()):
+            BaseService.session.add(word_info)
+        BaseService.session.commit()
+
+    # Everything that follows is private
```
Comment (mircealungu) [inline wordstats/cognate_info.py]: if it's private, the convention is to have it named with an underscore as prefix. e.g. `_print_stats`

File: wordstats/cognate_info.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    # generates candidates based on distance function func and word lists
+    def apply_distance_metric(self, wordlist1, wordlist2, func):
+
+        for w1 in wordlist1:
+            newlist = list(filter(lambda x: len(x)>0 , map(lambda w: w1 + ' ' + w if func(w1, w) else "", wordlist2)))
+            self.candidates.extend(newlist)
+
+    # note: only executable in terminal, reactive to one-key stroke
+    def start_quiz(self):
```
Comment (mircealungu) [inline wordstats/cognate_info.py]: one thing that would be good is to move all the stuff related to the quiz into a separate module. 

the cognates class can work independent of the way the blacklist / whitelists are generated.

the quiz is a different concern, and as we know, every class should encapsulate one single responsibility. 

agreed?

File: wordstats/word_distance.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
+import codecs
+import configparser
+from .cognate_files_path import *
+
+class WordDistance(object):
+    def __init__(self):
```
Comment (mircealungu) [inline wordstats/word_distance.py]: a class comment would help.
also explaining the meaning of the instance variables: replace_distance, add_distance, and threshold.

File: wordstats/cognate_db.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+class CognateCandidatesInfo(SimplifiedQuery, Base):
+    __tablename__ = 'candidates_info'
+    __table_args__ = {'mysql_collate': 'utf8_bin'}
+
+    id = Column(Integer, primary_key=True)
+    method = Column(String(255), nullable=False, index=True)
+
+    word_from = Column(String(255), nullable =False, index = True)
+    word_to = Column(String(255), nullable=False, index=True)
+    language_ids = Column(String(4), nullable =False, index = True)
```
Comment (mircealungu) [inline wordstats/cognate_db.py]: not all languages have two-letter codes: https://github.com/zeeguu-ecosystem/Zeeguu-Core/blob/master/zeeguu/model/language.py#L26

also, in the database, it makes little sense to have joined_language_ids. 
or
?

File: wordstats/edit_distance_absolute.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+            return True
+        else:
+            False
+
+
+    def initialize_from_config(self, config):
+        pass
+
+
+
```
Comment (mircealungu) [inline wordstats/edit_distance_absolute.py]: why so many whitelines here?

File: wordstats/edit_distance_overlap.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    def _edit_distance_overlap(self, word1:str, word2:str):
+
+        wordLongest, wordShortest = (word1, word2) if len(word1) >= len(word2) else (word2, word1)
+
+        overlap_size = 0
+        for i in range(len(wordShortest)):
+            for j in range(i + overlap_size, len(wordShortest)):
+                if wordLongest.find(wordShortest[i:(j + 1)]) >= 0:
+                    if (wordLongest == "het" and wordShortest == "he") or (wordShortest == "het" and wordLongest == "he"):
```
Comment (mircealungu) [inline wordstats/edit_distance_overlap.py]: hmm.
why do we hardcode these two words in here?

**Comment by joelgrondman** [comment]
The hardcoding was actually some quick debugging I forgot to remove, its
gone now.

About the database, what do you mean by joined_language_ids?
Should we create a database for each pair of languages instead?

**Comment by mircealungu** [comment]
>
> About the database, what do you mean by joined_language_ids?
> Should we create a database for each pair of languages instead?
>

What I meant is that it would make more sense to
have two columns instead of a single column with
two melded codes.

Or is there a strong reason for the one-single-column design?


>
> On Fri, Apr 27, 2018 at 2:28 PM, Mircea Filip Lungu <

**Comment by mircealungu** [comment]
@joelgrondman  - I see a lot of tests failing. at a closer look it seems that they have assertions that expect words in the top 1500. thus they can be fixed if i set `MAX_WORDS = 1500` in the config. 
however, the test cognates are failing in this case with the following error: 

`sqlalchemy.exc.IntegrityError: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely) (sqlite3.IntegrityError) UNIQUE constraint failed: candidates_info.word_from, candidates_info.word_to, candidates_info.languageFrom, candidates_info.languageTo, candidates_info.method [SQL: 'INSERT INTO candidates_info (method, word_from, word_to, "languageFrom", "languageTo") VALUES (?, ?, ?, ?, ?)'] [parameters: ('absolute', 'hier', 'hier', 'de', 'nl')]
`

can you investigate what's the problem?

**Comment by mircealungu** [comment]
more precisely CognateTests.testcachetodb fails with the previous error. 
the other tests are just running for a veeery long time.

File: wordstats/cognate_info.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        if output:
+            new_registry.print_load_stats(a, b)
+
+        return b - a
+
+    @classmethod
+    def _pprofile_load_from_file(cls, language_ids, method_name= "", output=False):
+
+        a = datetime.now()
+        new_registry = cls.load_from_file(languageFrom,languageTo, method_name= "")
```
Comment (mircealungu) [inline wordstats/cognate_info.py:185]: languageFrom languageTo are undefined at this point.

File: wordstats/tests/test_cognate.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+class CognateTests(TestCase):
+
+    @classmethod
+    def testload(self):
+
+        info = CognateInfo.load_from_path("de", "nl","edit_distance")
+
+
+    @classmethod
+    def testeditDistanceAbsolute(self):
```
Comment (mircealungu) [inline wordstats/tests/test_cognate.py:20]: testEditDistance with uppercase 'E'.
same later.

File: wordstats/tests/test_cognate.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+        cognateFM = CognateInfo("de", "nl","absolute")
+
+        distanceMetric = WordDistanceAbsolute()
+        distanceMetric.loadConfig("de", "nl", "absolute")
+        #if load fails create empty config file and warn user
+
+
+        cognateFM.apply_distance_metric(german, dutch, distanceMetric.edit_distance)
+        cognateFM.save_candidates()
```
Comment (mircealungu) [inline wordstats/tests/test_cognate.py:32]: i am a bit puzzled about these tests. 
what are they doing? 
they have no assertions in them.
every test should have one `assert` statement in it!

also, it seems that they modify the language_data folder. 
this is very risky. 

tests should never interfere with the actual data.

imagine that on the deployed server, somebody runs the
tests. that would destroy the valuable cognate information.

if i am right, we have to discuss about testing!

**Pull request closed without merging**